binary_search_tree/tree.py

```diff
@@ -14,42 +14,119 @@ class Tree:
+        return current
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)     
     def find(self, key):
-        pass
+        return self.find_node(self.root, key)
+    
+    def find_node(self,current,key):
+        if current == None:
+            return None
+        elif current.key == key:
+            return current.value
+        elif key < current.key:
+            return self.find_node(current.left, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def inorder(self):
-        pass
+        return self.find_node(current.right, key)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def inorder(self):
+        node_list = []
+        if not self.root:
+            return node_list
+        return self.inorder_traverse(self.root, node_list)
+    
+    def inorder_traverse(self, root, node_list):
+        if root:
+            self.inorder_traverse(root.left, node_list)
+            node_list.append({"key": root.key, "value": root.value})
+            self.inorder_traverse(root.right, node_list)
+            return node_list
+        
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
-
-    # Time Complexity: 
-    # Space Complexity:     
+        node_list = []
+        if not self.root:
+            return node_list
+        return self.preorder_traverse(self.root, node_list)
+    
+    def preorder_traverse(self, root, node_list):
+        if root:
+            node_list.append({"key": root.key, "value": root.value})
+            self.preorder_traverse(root.left, node_list)
+            self.preorder_traverse(root.right, node_list)
+            return node_list
+        
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def postorder(self):
-        pass
+        node_list = []
+        if not self.root:
+            return node_list
+        return self.postorder_traverse(self.root, node_list)
+    
+    def postorder_traverse(self, root, node_list):
+        if root:
+            self.postorder_traverse(root.left, node_list)
+            self.postorder_traverse(root.right, node_list)
+            node_list.append({"key": root.key, "value": root.value})
+            return node_list
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(h) (h is height of the tree)  
     def height(self):
-        pass
+        return self.get_height(self.root)
+    
+    def get_height(self, node):
+        if not node:
+            return 0
+        leftHeight = self.get_height(node.left)
+        rightHeight = self.get_height(node.right)
+        return max(leftHeight, rightHeight) + 1
+        
+        
 
 
 #   # Optional Method
-#   # Time Complexity: 
-#   # Space Complexity: 
+#   # Time Complexity: O(n)
+#   # Space Complexity: O(h) (h is height of the tree) 
     def bfs(self):
```

Comment:
👍 But I don't think you need to break this into two methods really.

Also the space complexity is O(n) because of the list returned.